src/stores/NotificationsStore.js

```diff
@@ -0,0 +1,280 @@
+  friendUserName: string,
+};
+
+export type LowKegLevelNotification = BaseNotification<
+  LowKegLevelNotificationExtraProps,
+>;
+
+export type NewAchievementNotification = BaseNotification<
+  NewAchievementNotificationExtraProps,
+>;
+
+export type NewFriendRequestNotification = BaseNotification<
+  NewFriendRequestNotificationExtraProps,
+>;
+
+export type Notification =
+  | LowKegLevelNotification
+  | NewAchievementNotification
+  | NewFriendRequestNotification;
+
+class NotificationsStore {
+  @observable _navigation: ?Navigation = null;
+  @observable
+  _notificationsByID: ObservableMap<BaseNotification<any>> = observable.map();
+
+  constructor() {
+    AppState.addEventListener('change', nextAppState => {
+      if (nextAppState === 'active') {
+        FCM.removeAllDeliveredNotifications();
+      }
+    });
+
+    FCM.on(FCMEvent.Notification, this._onRawNotification);
+
+    (async () => {
+      await this._rehydrateNotifications();
+
+      reaction(
+        () => this._notificationsByID.values(),
+        notifications => {
+          AsyncStorage.setItem(
+            NOTIFICATIONS_STORAGE_KEY,
+            JSON.stringify(notifications),
+          );
+        },
+        {
+          compareStructural: true,
+        },
+      );
+    })();
+  }
+
+  @computed
+  get notifications(): Array<Notification> {
+    return this._notificationsByID
+      .values()
+      .sort((a, b): number => new Date(b.date) - new Date(a.date));
+  }
+
+  @computed
+  get unreadCount(): number {
+    return this._notificationsByID
+      .values()
+      .filter(({ isRead }: Notifications) => !isRead).length;
+  }
+
+  @computed
+  get hasUnread(): boolean {
+    return this.unreadCount > 0;
+  }
+
+  @computed
+  get _isInitialized(): boolean {
+    return !!this._navigation && AuthStore.isAuthorized;
+  }
+
+  handleInitialNotification = async () => {
+    // handle case when the app is killed and we launch it by clicking
+    // on incoming notifications.
+    // the timeout is a workaround to avoid the warning:
+    // "Tried to use permissions API while not attached to an Activity"
+    // https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/permissions/PermissionsModule.java
+    // https://github.com/facebook/react-native/issues/13439
+    // it appears sometimes because of race conditions between native/js thread probably.
+    // they use timeout even in react-native-fcm example app :(
+    // https://github.com/evollu/react-native-fcm/blob/master/Examples/simple-fcm-client/app/App.js#L44
+    setTimeout(async () => {
+      const initialNotification = await FCM.getInitialNotification();
+      initialNotification && this._onRawNotification(initialNotification);
+    }, 500);
+  };
+
+  registerToken = async () => {
+    const fcmToken = await FCM.getFCMToken();
+    const deviceUniqueID = DeviceInfo.getUniqueID();
+
+    await fetchJSON(`${BASE_PUSH_URL}/register`, {
```

Comment:
I'm going to rethink the API a bit (next week). We'll need a `register` just to connect to FCM but we'll need the ability to unsubscribe from specific entities (taps/friends) and more in the future. This here should stay the same though.

Reply:
yeah. I remember that we need taps opt-out.